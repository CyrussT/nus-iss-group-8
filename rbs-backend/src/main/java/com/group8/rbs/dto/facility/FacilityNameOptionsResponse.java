package com.group8.rbs.dto.facility;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class FacilityNameOptionsResponse {
    private Long facilityTypeId;
    private String facilityType;
}
