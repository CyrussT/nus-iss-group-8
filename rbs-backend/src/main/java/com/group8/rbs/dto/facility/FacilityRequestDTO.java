package com.group8.rbs.dto.facility;

import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Data
@Builder
@NoArgsConstructor 
@AllArgsConstructor
public class FacilityRequestDTO {

    @NotBlank(message = "Resource type cannot be empty")
    private String resourceType;

    @NotBlank(message = "Resource name cannot be empty")
    private String resourceName;

    @NotBlank(message = "Location cannot be empty")
    private String location;

    @NotNull(message = "Capacity is required")
    @Min(value = 1, message = "Capacity must be at least 1")
    private Integer capacity;
}