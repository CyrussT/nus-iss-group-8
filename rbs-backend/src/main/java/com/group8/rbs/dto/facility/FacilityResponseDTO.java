package com.group8.rbs.dto.facility;

import java.util.List;

import com.group8.rbs.dto.booking.BookingResponseDTO;

import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class FacilityResponseDTO {
    private Long facilityId;
    private Long resourceTypeId;
    private String resourceName;
    private String location;
    private Integer capacity;
    private String message;
    private List<BookingResponseDTO> bookings;
}
