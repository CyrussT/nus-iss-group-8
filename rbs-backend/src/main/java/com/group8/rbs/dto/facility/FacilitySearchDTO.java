package com.group8.rbs.dto.facility;

import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class FacilitySearchDTO {
    private Long resourceTypeId;
    private String resourceName;
    private String location;
    private Integer capacity;
}
