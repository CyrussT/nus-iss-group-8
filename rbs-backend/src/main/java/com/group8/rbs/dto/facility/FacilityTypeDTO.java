package com.group8.rbs.dto.facility;


import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class FacilityTypeDTO {
    private Long id;
    private String name;
}