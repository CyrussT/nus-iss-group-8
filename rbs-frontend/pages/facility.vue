<script setup lang="ts">
definePageMeta({
  middleware: ['auth', 'admin']
});

import { useRouter } from "vue-router";
import { computed, onMounted, ref } from "vue";
import type { Facility } from "@/composables/useFacility";
import { useFacility } from "@/composables/useFacility";
import axios from "axios";


const router = useRouter();
const isModalOpen = ref(false);
const isEditing = ref(false);

definePageMeta({
  middleware: ['auth', 'admin']
})

const {
  searchQuery,
  fetchFacilities,
  resetSearch,
  facility,
  resetFacility,
  facilities,
  loading,
  currentPage,
  totalItems,
  pageSize,
  fetchResourceTypes,
  resourceTypeOptions,
} = useFacility();


const openModal = (facilityData: Partial<Facility> | null = null) => {
  if (facilityData && typeof facilityData === "object") {
    facility.value = {
      facilityId: facilityData.facilityId ?? undefined,
      resourceTypeId: facilityData.resourceTypeId ?? undefined,
      resourceName: facilityData.resourceName ?? "",
      location: facilityData.location ?? "",
      capacity: facilityData.capacity ?? 1,
      bookings: facilityData.bookings ?? [],
    };
    isEditing.value = true;
  } else {
    resetFacility();
    isEditing.value = false;
  }
  isModalOpen.value = true;
};

const markAsMaintenance = (row: any) => {
  console.log(`Maintenance mode triggered for facility: ${row.facilityId}`);
  alert("Maintenance mode triggered for facility: " + row.facilityId);
};

const closeModal = () => {
  isModalOpen.value = false;
  resetFacility();
};

const viewFacility = (facilityId: number) => {
  router.push(`/facility-details/${facilityId}`);
};

const saveFacility = async () => {
  try {
    if (isEditing.value) {
      await axios.put(`http://localhost:8080/api/facilities/update/${facility.value.facilityId}`, facility.value);
      alert("Facility Updated Successfully!");
    } else {
      await axios.post("http://localhost:8080/api/facilities/create", facility.value);
      alert("Facility Created Successfully!");
    }
    closeModal();
    fetchFacilities();
  } catch (error) {
    console.error("Error saving facility:", error);
    alert("Failed to save facility. Please try again.");
  }
};



onMounted(async () => {
  await fetchResourceTypes();
  console.log("resourceTypeOptions:", resourceTypeOptions.value); // Add this line
  fetchFacilities();
});

</script>

<template>
  <div class="p-8">
    <h1 class="text-2xl font-bold mb-4">Facility Management</h1>

    <div class="grid grid-cols-2 gap-4">


      <UInputMenu v-model="searchQuery.resourceTypeId" :options="resourceTypeOptions" option-attribute="name"
        value-attribute="id" placeholder="Type or select resource type" size="md" class="w-full"
        clearable />

      <UInput v-model="searchQuery.resourceName" placeholder="Resource Name" />
      <UInput v-model="searchQuery.location" placeholder="Location" />
      <UInput v-model="searchQuery.capacity" type="number" placeholder="Capacity" />

      <div class="col-span-2 flex justify-end gap-2">

        <UButton @click="() => { currentPage = 1; fetchFacilities(); }" color="blue" variant="solid"
          icon="i-ic:baseline-search" label="Search" class="px-4 py-2 gap-2" />


        <UButton @click="resetSearch" color="gray" variant="solid" icon="i-ic:round-restart-alt" label="Reset"
          class="px-4 py-2 gap-2 hover:bg-red-500" />

        <UButton @click="openModal()" color="green" variant="solid" icon="i-ic:baseline-plus" label="Add Facility"
          class="px-4 py-2 gap-2" />

      </div>
    </div>

    <UCard class="w-full max-w-8xl p-6 shadow-lg bg-white">
      <div class="mt-6 overflow-x-auto">
        <UTable :rows="facilities" :loading="loading" :columns="[
          { key: 'sn', label: 'SN', sortable: false },
          { key: 'resourceTypeName', label: 'Resource Type', sortable: true },
          { key: 'resourceName', label: 'Name', sortable: true },
          { key: 'location', label: 'Location', sortable: true },
          { key: 'capacity', label: 'Capacity', sortable: true },
          { key: 'actions', label: 'Actions', class: 'text-center' }
        ]">
          <template #sn-data="{ index }">
            {{ index + 1 + (currentPage - 1) * pageSize }}
          </template>

          <template #actions-data="{ row }">
            <div class="flex justify-center gap-2">

              <UButton @click="viewFacility(row.facilityId)" color="blue" variant="solid" icon="i-heroicons-eye"
                label="View" />

              <UButton @click="openModal(row)" color="yellow" variant="solid" icon="i-heroicons-pencil" label="Edit"
                class="px-3 py-1 gap-2" />


              <UButton @click="markAsMaintenance(row)" color="red" variant="solid" icon="i-heroicons-wrench"
                label="Maintenance" class="px-3 py-1 gap-2" />

            </div>
          </template>
        </UTable>
      </div>

      <div class="mt-4 flex justify-center">
        <UPagination v-model="currentPage" :max="5" :total="totalItems" @update:model-value="fetchFacilities" show-last
          show-first />
      </div>
    </UCard>
  </div>
  <UModal v-model="isModalOpen">
    <UCard class="p-9 max-w-lg ">
      <button @click="closeModal" class="absolute top-4 right-4 text-gray-500 hover:text-gray-700">
        <UIcon name="i-heroicons-x-mark" class="w-6 h-6" />
      </button>

      <h2 class="text-xl font-bold mb-4 text-center">
        {{ isEditing ? "Edit Facility" : "Add Facility" }}
      </h2>
      <div class="grid grid-cols-3 gap-4 items-center">
        <label class="text-gray-700 font-medium col-span-1">Resource Type:</label>

          
      <UInputMenu v-model="facility.resourceTypeId" :options="resourceTypeOptions" option-attribute="name"
        value-attribute="id" placeholder="Type or select resource type" size="md" class="col-span-2 w-full"
        clearable />


        <label class="text-gray-700 font-medium col-span-1">Resource Name:</label>
        <UInput v-model="facility.resourceName" class="col-span-2 w-full" />

        <label class="text-gray-700 font-medium col-span-1">Location:</label>
        <UInput v-model="facility.location" class="col-span-2 w-full" />

        <label class="text-gray-700 font-medium col-span-1">Capacity:</label>
        <UInput v-model.number="facility.capacity" type="number" class="col-span-2 w-full" />
      </div>


      <div class="mt-9 flex justify-end gap-3">
        <button @click="closeModal" class="bg-gray-400 text-white px-4 py-2 rounded hover:bg-gray-600">
          Cancel
        </button>
        <button @click="saveFacility" class="bg-green-500 text-white px-4 py-2 rounded hover:bg-green-700">
          {{ isEditing ? "Update Facility" : "Save Facility" }}
        </button>
      </div>
    </UCard>
  </UModal>
</template>
