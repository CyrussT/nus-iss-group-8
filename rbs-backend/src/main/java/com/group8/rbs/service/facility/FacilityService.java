package com.group8.rbs.service.facility;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.RequestBody;

import com.group8.rbs.dto.facility.FacilityRequestDTO;
import com.group8.rbs.dto.facility.FacilityResponseDTO;
import com.group8.rbs.dto.facility.FacilitySearchDTO;
import com.group8.rbs.entities.Facility;
import com.group8.rbs.exception.FacilityException;
import com.group8.rbs.mapper.FacilityMapper;
import com.group8.rbs.repository.FacilityRepository;
import com.group8.rbs.specification.FacilitySpecification;

import jakarta.validation.Valid;


@Service
public class FacilityService {
    private final FacilityRepository facilityRepository;
    private final FacilityMapper facilityMapper;
    private static final Logger logger = LoggerFactory.getLogger(FacilityService.class);

    @Autowired
    public FacilityService(FacilityRepository facilityRepository, FacilityMapper facilityMapper) {
        this.facilityRepository = facilityRepository;
        this.facilityMapper = facilityMapper;
    }

    // Create facility
    public FacilityResponseDTO createFacility(FacilityRequestDTO facilityRequestDTO) {
        Optional<Facility> existingFacility = facilityRepository.findByResourceNameAndLocation(
            facilityRequestDTO.getResourceName(), facilityRequestDTO.getLocation()
        );

        if (existingFacility.isPresent()) {
            throw new FacilityException.FacilityAlreadyExistsException("Facility with this name and location already exists.");
        }

        Facility facility = facilityMapper.toEntity(facilityRequestDTO);
        Facility savedFacility = facilityRepository.save(facility);

        FacilityResponseDTO responseDTO = facilityMapper.toDTO(savedFacility);
        responseDTO.setMessage("Facility created successfully!");
        return responseDTO;
    }

    public List<FacilityResponseDTO> getAllFacilities() {
        return facilityRepository.findAll().stream()
                .map(facilityMapper::toDTO)
                .collect(Collectors.toList());
    }


    public FacilityResponseDTO getFacilityById(Long id) {
        Facility facility = facilityRepository.findById(id)
                .orElseThrow(() -> new FacilityException.FacilityNotFoundException("Facility not found"));
        return facilityMapper.toDTO(facility);
    }


    public FacilityResponseDTO updateFacility(Long id, FacilityRequestDTO facilityRequestDTO) {
        return facilityRepository.findById(id).map(facility -> {
            facility.setResourceType(facilityRequestDTO.getResourceType());
            facility.setResourceName(facilityRequestDTO.getResourceName());
            facility.setLocation(facilityRequestDTO.getLocation());
            facility.setCapacity(facilityRequestDTO.getCapacity());
            Facility updatedFacility = facilityRepository.save(facility);
            return facilityMapper.toDTO(updatedFacility);
        }).orElseThrow(() -> new RuntimeException("Facility not found"));
    }


    public void deleteFacility(Long id) {
        Facility facility = facilityRepository.findById(id)
                .orElseThrow(() -> new FacilityException.FacilityNotFoundException("Facility not found"));
        facilityRepository.delete(facility);
    }

    public Page<FacilityResponseDTO> searchFacilities(FacilitySearchDTO searchDTO, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        
        Page<Facility> facilityPage = facilityRepository.findAll(FacilitySpecification.searchFacilities(searchDTO), pageable);

        return facilityPage.map(facilityMapper::toDTO);
    }

    
}
