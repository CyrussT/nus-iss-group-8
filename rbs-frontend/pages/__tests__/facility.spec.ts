import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import facility from '../facility.vue';

describe('facility', () => {
  it('renders correctly', () => {
    const wrapper = mount(facility, {
      global: {
        stubs: {
          UButton: true,
          UIcon: true
        }
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
