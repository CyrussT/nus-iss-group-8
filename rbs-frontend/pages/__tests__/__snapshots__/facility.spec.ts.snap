// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`facility > renders correctly 1`] = `
"<div data-v-629372b2="" class="p-8 dark:bg-gray-900 dark:text-white">
  <h1 data-v-629372b2="" class="text-2xl font-bold mb-4 dark:text-white">Facility Management</h1>
  <div data-v-629372b2="" class="grid grid-cols-2 gap-4">
    <div class="relative w-full dark:bg-gray-800 dark:text-white dark:border-gray-700" data-headlessui-state="">
      <div class="flex items-center w-full"><input value="" id="headlessui-combobox-input-v-0" aria-expanded="false" aria-autocomplete="list" role="combobox" type="text" tabindex="0" data-headlessui-state="" placeholder="Type or select resource type" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-3 py-2 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400 pe-10" autocomplete="off" clearable="">
        <!--v-if--><button id="headlessui-combobox-button-v-1" type="button" tabindex="-1" aria-haspopup="listbox" aria-expanded="false" data-headlessui-state="" class="absolute inset-y-0 end-0 flex items-center px-3">
          <u-icon-stub name="i-heroicons-chevron-down-20-solid" class="flex-shrink-0 text-gray-400 dark:text-gray-500 h-5 w-5"></u-icon-stub>
        </button>
      </div>
      <!--v-if-->
    </div>
    <div data-v-629372b2="" class="relative dark:bg-gray-800 dark:text-white dark:border-gray-700"><input type="text" placeholder="Resource Name" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400" value="">
      <!--v-if-->
      <!--v-if-->
    </div>
    <div data-v-629372b2="" class="relative dark:bg-gray-800 dark:text-white dark:border-gray-700"><input type="text" placeholder="Location" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400" value="">
      <!--v-if-->
      <!--v-if-->
    </div>
    <div data-v-629372b2="" class="relative dark:bg-gray-800 dark:text-white dark:border-gray-700"><input type="number" placeholder="Capacity" class="relative block w-full disabled:cursor-not-allowed disabled:opacity-75 focus:outline-none border-0 form-input rounded-md placeholder-gray-400 dark:placeholder-gray-500 text-sm px-2.5 py-1.5 shadow-sm bg-white dark:bg-gray-900 text-gray-900 dark:text-white ring-1 ring-inset ring-gray-300 dark:ring-gray-700 focus:ring-2 focus:ring-primary-500 dark:focus:ring-primary-400" value="">
      <!--v-if-->
      <!--v-if-->
    </div>
    <div data-v-629372b2="" class="col-span-2 flex justify-end gap-2"><button data-v-629372b2="" type="button" class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm shadow-sm text-white dark:text-gray-900 bg-blue-500 hover:bg-blue-600 disabled:bg-blue-500 aria-disabled:bg-blue-500 dark:bg-blue-400 dark:hover:bg-blue-500 dark:disabled:bg-blue-400 dark:aria-disabled:bg-blue-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-500 dark:focus-visible:outline-blue-400 inline-flex items-center px-4 py-2 gap-2">
        <u-icon-stub name="i-heroicons-outline:magnifying-glass" class="flex-shrink-0 h-5 w-5" aria-hidden="true"></u-icon-stub><span class="">Search</span>
        <!--v-if-->
      </button><button data-v-629372b2="" type="button" class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm shadow-sm ring-1 ring-inset ring-gray-300 dark:ring-gray-700 text-gray-700 dark:text-gray-200 bg-gray-50 disabled:bg-gray-50 aria-disabled:bg-gray-50 dark:bg-gray-800 dark:disabled:bg-gray-800 dark:aria-disabled:bg-gray-800 focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center px-4 py-2 gap-2 hover:bg-red-500 dark:hover:bg-red-600">
        <u-icon-stub name="i-heroicons-outline:arrow-path" class="flex-shrink-0 h-5 w-5" aria-hidden="true"></u-icon-stub><span class="">Reset</span>
        <!--v-if-->
      </button><button data-v-629372b2="" type="button" class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm shadow-sm text-white dark:text-gray-900 bg-green-500 hover:bg-green-600 disabled:bg-green-500 aria-disabled:bg-green-500 dark:bg-green-400 dark:hover:bg-green-500 dark:disabled:bg-green-400 dark:aria-disabled:bg-green-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-green-500 dark:focus-visible:outline-green-400 inline-flex items-center px-4 py-2 gap-2">
        <u-icon-stub name="i-heroicons-outline:plus" class="flex-shrink-0 h-5 w-5" aria-hidden="true"></u-icon-stub><span class="">Add Facility</span>
        <!--v-if-->
      </button></div>
  </div>
  <div data-v-629372b2="" class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 w-full max-w-8xl p-6 shadow-lg bg-white dark:bg-gray-800 dark:border-gray-700 mt-4">
    <!--v-if-->
    <div class="px-4 py-5 sm:p-6">
      <div data-v-629372b2="" class="mt-6 overflow-x-auto">
        <div data-v-629372b2="" class="relative overflow-x-auto dark:text-gray-200">
          <table class="min-w-full table-fixed divide-y divide-gray-300 dark:divide-gray-700">
            <!--v-if-->
            <thead class="relative">
              <tr class="">
                <!--v-if-->
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm"><span>SN</span></th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm" aria-sort="none">
                  <u-button-stub type="button" block="false" label="Resource Type" loading="false" disabled="false" padded="true" size="sm" color="gray" variant="ghost" icon="i-heroicons-arrows-up-down-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="true" leading="false" square="true" truncate="false" class="-m-1.5" ui="[object Object]"></u-button-stub>
                </th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm" aria-sort="none">
                  <u-button-stub type="button" block="false" label="Name" loading="false" disabled="false" padded="true" size="sm" color="gray" variant="ghost" icon="i-heroicons-arrows-up-down-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="true" leading="false" square="true" truncate="false" class="-m-1.5" ui="[object Object]"></u-button-stub>
                </th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm" aria-sort="none">
                  <u-button-stub type="button" block="false" label="Location" loading="false" disabled="false" padded="true" size="sm" color="gray" variant="ghost" icon="i-heroicons-arrows-up-down-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="true" leading="false" square="true" truncate="false" class="-m-1.5" ui="[object Object]"></u-button-stub>
                </th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm" aria-sort="none">
                  <u-button-stub type="button" block="false" label="Capacity" loading="false" disabled="false" padded="true" size="sm" color="gray" variant="ghost" icon="i-heroicons-arrows-up-down-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="true" leading="false" square="true" truncate="false" class="-m-1.5" ui="[object Object]"></u-button-stub>
                </th>
                <th scope="col" class="text-left rtl:text-right px-4 py-3.5 text-gray-900 dark:text-white font-semibold text-sm text-center"><span>Actions</span></th>
              </tr>
              <tr>
                <td colspan="0" class="absolute inset-x-0 -bottom-[0.5px] p-0">
                  <div data-v-4e5d44bb="" class="w-full flex flex-col gap-2" role="progressbar">
                    <!--v-if--><progress data-v-4e5d44bb="" class="block appearance-none border-none overflow-hidden w-full [&amp;::-webkit-progress-bar]:w-full h-px rounded-full [&amp;::-webkit-progress-bar]:rounded-full [&amp;::-webkit-progress-bar]:bg-gray-200 [&amp;::-webkit-progress-bar]:dark:bg-gray-700 [@supports(selector(&amp;::-moz-progress-bar))]:bg-gray-200 [@supports(selector(&amp;::-moz-progress-bar))]:dark:bg-gray-700 [&amp;::-webkit-progress-value]:rounded-full [&amp;::-webkit-progress-value]:transition-all [&amp;::-webkit-progress-value]:ease-in-out [&amp;::-moz-progress-bar]:rounded-full text-primary-500 dark:text-primary-400 [&amp;::-webkit-progress-value]:bg-current [&amp;::-moz-progress-bar]:bg-current indeterminate:relative indeterminate:after:rounded-full [&amp;:indeterminate::-webkit-progress-value]:rounded-full [&amp;:indeterminate::-moz-progress-bar]:rounded-full bar-animation-carousel"></progress>
                    <!--v-if-->
                  </div>
                </td>
              </tr>
            </thead>
            <tbody class="divide-y divide-gray-200 dark:divide-gray-800">
              <tr>
                <td colspan="6">
                  <div class="flex flex-col items-center justify-center flex-1 px-6 py-14 sm:px-14">
                    <u-icon-stub name="i-heroicons-arrow-path-20-solid" class="w-6 h-6 mx-auto text-gray-400 dark:text-gray-500 mb-4 animate-spin" aria-hidden="true"></u-icon-stub>
                    <p class="text-sm text-center text-gray-900 dark:text-white">Loading...</p>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div data-v-629372b2="" class="mt-4 flex justify-center">
        <div data-v-629372b2="" class="flex items-center -space-x-px dark:text-white">
          <u-button-stub type="button" block="false" loading="false" disabled="true" padded="true" size="sm" color="white" variant="solid" icon="i-heroicons-chevron-double-left-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="false" leading="false" square="false" truncate="false" class="first:rounded-s-md last:rounded-e-md rtl:[&amp;_span:first-child]:rotate-180" ui="[object Object]" aria-label="First"></u-button-stub>
          <u-button-stub type="button" block="false" loading="false" disabled="true" padded="true" size="sm" color="white" variant="solid" icon="i-heroicons-chevron-left-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="false" leading="false" square="false" truncate="false" class="first:rounded-s-md last:rounded-e-md rtl:[&amp;_span:first-child]:rotate-180" ui="[object Object]" aria-label="Prev"></u-button-stub>
          <u-button-stub type="button" block="false" loading="false" disabled="true" padded="true" size="sm" color="white" variant="solid" icon="i-heroicons-chevron-right-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="false" leading="false" square="false" truncate="false" class="first:rounded-s-md last:rounded-e-md rtl:[&amp;_span:last-child]:rotate-180" ui="[object Object]" aria-label="Next"></u-button-stub>
          <u-button-stub type="button" block="false" loading="false" disabled="true" padded="true" size="sm" color="white" variant="solid" icon="i-heroicons-chevron-double-right-20-solid" loadingicon="i-heroicons-arrow-path-20-solid" trailing="false" leading="false" square="false" truncate="false" class="first:rounded-s-md last:rounded-e-md rtl:[&amp;_span:last-child]:rotate-180" ui="[object Object]" aria-label="Last"></u-button-stub>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>
<!-- Facility Modal -->
<!-- Maintenance Modal -->
<!-- Early Release Confirmation Modal -->"
`;
